atlasaction/action_test.go

```diff
@@ -87,6 +93,234 @@ func TestMigrateApply(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "broken-driver://")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `unknown driver "broken-driver"`)
+	})
+	t.Run("invalid tag", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.setInput("tag", "invalid-character@")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `tag must be lowercase alphanumeric`)
+	})
+	t.Run("tag", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.setInput("tag", "valid-tag-123")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+	})
+	t.Run("e2e test", func(t *testing.T) {
+		var syncDirCalls, pushDirCalls atomicInt
+
+		tt := newHttpT(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			query := graphQLQuery{}
+			err := json.NewDecoder(r.Body).Decode(&query)
+			require.NoError(t, err)
+			switch {
+			case strings.Contains(query.Query, "syncDir"):
+				syncDirCalls.inc()
+				require.NoError(t, json.Unmarshal(query.Variables, &query.SyncDir))
+				require.Equal(t, "test-dir", query.SyncDir.Slug)
+				expected := &ContextInput{
+					Repo:   "repository",
+					Path:   "file://testdata/migrations",
+					Branch: "testing-branch",
+					Commit: "sha1234",
+					URL:    "",
+				}
+				require.Equal(t, expected, query.SyncDir.Context)
+			case strings.Contains(query.Query, "pushDir"):
+				pushDirCalls.inc()
+				require.NoError(t, json.Unmarshal(query.Variables, &query.PushDir))
+				require.Equal(t, query.PushDir.Tag, "sha1234")
+				require.Equal(t, query.PushDir.Slug, "test-dir")
+				fmt.Fprint(w, `{"data":{"pushDir":{"url":"https://some-org.atlasgo.cloud/dirs/314159/tags/12345"}}}`)
+			}
+		}))
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.env["GITHUB_REPOSITORY"] = "repository"
+		tt.env["GITHUB_REF_NAME"] = "testing-branch"
+		tt.env["GITHUB_SHA"] = "sha1234"
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+		require.Equal(t, syncDirCalls.get(), int32(1))
+		require.Equal(t, pushDirCalls.get(), int32(1))
+		require.NoError(t, err)
+		outputs, _ := tt.outputs()
+		url := outputs["url"]
+		require.Equal(t, "https://some-org.atlasgo.cloud/dirs/314159/tags/12345", url)
+	})
+}
+
+type (
+	httpTest struct {
+		*test
+		srv        *httptest.Server
+		cloudToken string
+	}
```

Comment:
It looks like this type is not needed and can be defined as a function scoped to the specific test. See my comment above

Reply:
@a8m If I do that, how do I use these types from other functions (i.e `newHttpT`)?

Reply:
The usage of this API can be done like this:
```
token := value
url := mockCloud(tt, token, handler)
generateHCL(tt, url, token)
```

Reply:
Also, see https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Reply:
I hope I understood you correctly

https://github.com/ariga/atlas-action/pull/69/commits/2843a4b02bae755aaec80b8d63d6dba94631a323